overlord/servicestate/quota_handlers.go

```diff
@@ -823,7 +853,77 @@ func validateSnapForAddingToGroup(st *state.State, snaps []string, group string,
 			}
 		}
 	}
+	return nil
+}
+
+// splitSnapServiceName splits and verifies the snap service reference
+// taken in by the frontend. It expects the format snap.service
+func splitSnapServiceName(name string) (string, string, error) {
+	parts := strings.Split(name, ".")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid snap service: %s", name)
+	}
+	if err := naming.ValidateSnap(parts[0]); err != nil {
+		return "", "", err
+	}
+	return parts[0], parts[1], naming.ValidateApp(parts[1])
+}
+
+// ensureAppReferenceIsService returns whether the service referred to in the
+// snap is actually a service.
+func ensureAppReferenceIsService(st *state.State, snap, service string) error {
+	snapInfo, err := snapstate.CurrentInfo(st, snap)
+	if err != nil {
+		return err
+	}
+
+	appInfo, ok := snapInfo.Apps[service]
+	if !ok {
+		return fmt.Errorf("invalid service %q", service)
+	}
+	if !appInfo.IsService() {
+		return fmt.Errorf("%q is not a service", service)
+	}
+	return nil
+}
+
+// groupHasSubGroups returns true if the group has no sub-groups or it doesn't exist,
+// otherwise turns false.
+func groupHasSubGroups(group string, allGrps map[string]*quota.Group) bool {
+	grp, ok := allGrps[group]
+	if ok && len(grp.SubGroups) != 0 {
+		return false
+	}
+	return true
+}
+
+// validateSnapServicesForAddingToGroup verifies that the given services can be added
+// to the provided group. Services come in the format of snap.service, and we do
+// the split and parsing in this function. We need to make sure that the snaps referred
+// are related to the current group, and we make sure that the services are valid service names.
+//
+// We do not allow mixing services and sub-groups. A quota group of service must be
+// the final leaf node in the quota group tree.
+// State must locked prior to calling this function
+func validateSnapServicesForAddingToGroup(st *state.State, services []string, group string, parentGroup *quota.Group, allGrps map[string]*quota.Group) error {
+	if len(services) > 0 {
+		if !groupHasSubGroups(group, allGrps) {
+			return fmt.Errorf("cannot mix services and sub groups in the group %q", group)
+		}
+	}
 
+	for _, name := range services {
+		snap, service, err := splitSnapServiceName(name)
+		if err != nil {
+			return err
+		}
+		if err = ensureAppReferenceIsService(st, snap, service); err != nil {
+			return fmt.Errorf("cannot use snap service in group %q: %v", group, err)
+		}
+		if parentGroup != nil && !parentGroup.IsSnapRelated(snap) {
```

Comment:
it's actually an error in itself here, if parentGroup is nil, no? but it seems the code as is would add services to a top-group?

Reply:
Yea you are absolutely right, but points are really good. I missed that one. Thanks for pointing this out!